### PR TITLE
Editor: Add global styles data to editor settings

### LINKED ISSUE #49278: Styles: show a hierarchy of styles clearly (theme, user, global unspecific, global specific, local, etc)
### When no styles are detected
No styles detected

When styles cannot be detected (from a style provider like global styles / theme.json) no controls are displayed. This eliminates the confusion between undetectable inherited styles (e.g. custom css, browser styles) and empty/unset controls. See #43082 for additional context. 

### Working globally
Working globally

Respecting the convention above, panels are empty until a value is set (either in the UI, or via theme.json).

### Working locally
When working locally there are three states to consider:

**Inherited values**
Working locally

If a value is inherited (e.g. from global styles) the control displays that value, and the label has a purple accent denoting the global context. Hovering/focussing the label reveals the source in a tooltip. The control cannot be removed because if there's no local setting, the inherited value will be used.

**Local overrides**
Local overrides

If an inherited value is changed then a blue dot is appended to the label. Clicking the dot reveals a menu where the user can reset to the global default, or push the value upstream. Some controls (like background image, see below) may need an additional action to 'unset' the value locally.

**Local styles**
Styles can be added to a block in the local context that have no inherited value. The manifestation of this in the UI is essentially the same as the "Working globally" case above:

Working globally

With one caveat: If a global style for this property is later added, the local control would then exhibit the blue dot. 

### Examples
In addition to the block-spacing example above, here are a couple others that have come up in other issues recently. In each case the edit context is local to a document:

**Lightbox**
Lightbox

**Background image**
Background image

### Breakdown

This doesn't all need to be implemented at once. The work could be broken down into the following steps:

- [ ] https://github.com/WordPress/gutenberg/issues/37752
- [ ] If local styles override global then indicate that change with a dot which serves as a Reset affordance.
- [ ] Highlight where those styles are coming from.
- [ ] Implement make default/push to global functionality.

### Important note

The purple accents used to indicate inherited styles in the mockups above could be a bit heavy-handed in practise. We should continue to explore other design solutions for this.

<details><summary>Original issue</summary>Knowing what style is affecting something on a theme is difficult (is it a global element? is it coming from a block type? is it local to the template?). This has shown up in a few areas already, from Custom CSS https://github.com/WordPress/gutenberg/issues/47946 to the general experience of conflicting styles https://github.com/WordPress/gutenberg/issues/45086 https://github.com/WordPress/gutenberg/issues/44931. To improve this, we should find a way to show a hierarchy of styles clearly (theme, user, global unspecific, global specific, local, etc). While building this, we should also consider ways to push or transfer something to the right level when working through a design, which relates to https://github.com/WordPress/gutenberg/issues/44931

Image

In most cases understanding this style hierarchy is not crucial, as you can accomplish most of what you need to without the visualization. You go to global styles to style site wide, you edit pages to style locally. Nevertheless for more complex sessions, it can be useful to surface, but due to the rarity of needing to surface this, the footprint and UI weight of any visualization feature needs to be small. 

Here's a take that makes the design tool _label_ clickable to reveal the inheritance:

Image

* Click a label to see the inheritance
* If the popover has a link, that link should serve as a deep-link to the origin of the inheritance

For this concept to work, the Color flyout has received its own label.

Thoughts?</details>

### PR DESCRIPTION
**Pull request by michalczaplinski**
<details>
<summary>Original PR Description</summary>

## What?

This PR introduces a new `useGlobalStylesData()` hook in the `editor` package that gets the user global styles. Those user styles are then added to the block editor settings.

## Why?

Second attempt at building a foundation to fix #49278 and #43082.

The first attempt (https://github.com/WordPress/gutenberg/pull/55952) is probably not the right approach.

## How?

We add the "uncompiled" global styles to the editor settings.

</details>

## What?

Adds merged global styles data to the block editor settings.

## Why?

The global styles data is required in both editors for two main areas of ongoing work:
1. Displaying style inheritance
2. Section styling (a.k.a extended block style variations)

## How?

- A new hook, `useGlobalStylesData` has been added to the editor package but **not** exported as a public API.
- The above hook is used to collect and merge the global styles data from different origins i.e. user, base, theme etc.
- The merged global styles data is added to the block editor settings under a private key
- The approach to merged global styles data follows closely to the approach of the global styles provider in the site editor

## Testing

As this PR adds the data to a private key within the block editor settings, it can't easily be inspected via Redux dev tools.

The following snippet will log out the global styles data via a `console.debug()` call when there is a block with an element style applied. From there you can easily confirm the correct data. Bonus points if you make tweaks in theme.json and global styles, then reconfirm that all gets merged properly.

<details>
<summary>Test snippet</summary>

```diff
diff --git a/packages/block-editor/src/hooks/style.js b/packages/block-editor/src/hooks/style.js
index dad62bc059..49581878ac 100644
--- a/packages/block-editor/src/hooks/style.js
+++ b/packages/block-editor/src/hooks/style.js
@@ -1,6 +1,7 @@
 /**
  * WordPress dependencies
  */
+import { useSelect } from '@wordpress/data';
 import { useMemo } from '@wordpress/element';
 import { addFilter } from '@wordpress/hooks';
 import {
@@ -14,6 +15,8 @@ import { getCSSRules, compileCSS } from '@wordpress/style-engine';
 /**
  * Internal dependencies
  */
+import { store as blockEditorStore } from '../store';
+import { globalStylesDataKey } from '../store/private-keys';
 import { BACKGROUND_SUPPORT_KEY, BackgroundImagePanel } from './background';
 import { BORDER_SUPPORT_KEY, BorderPanel, SHADOW_SUPPORT_KEY } from './border';
 import { COLOR_SUPPORT_KEY, ColorEdit } from './color';
@@ -373,6 +376,12 @@ function useBlockProps( { name, style } ) {
 		useBlockProps
 	) }`;
 
+	const globalStyles = useSelect( ( select ) => {
+		const settings = select( blockEditorStore ).getSettings();
+		return settings[ globalStylesDataKey ];
+	}, [] );
+	console.debug( globalStyles );
+
 	// The .editor-styles-wrapper selector is required on elements styles. As it is
 	// added to all other editor styles, not providing it causes reset and global
 	// styles to override element styles because of higher specificity.

```
</details>

### CODE REVIEW

File: packages/edit-post/src/editor.js [hunk 1]
```diff
@@ -70,14 +71,24 @@ function Editor( {
...
 	);
 
+	// get the styles from the global styles
+	const { styles } = useStyles();
+
 	const editorSettings = useMemo(
 		() => ( {
 			...settings,
+			// TODO: This key should have a better name
+			__globalStyles: styles,
```
Comment (youknowriad) [inline packages/edit-post/src/editor.js]: It's not entirely clear to me why this is passed down from "edit-post" rather than just added automatically from `use-block-editor-settings.js` (directly in the editor package)

Comment (youknowriad) [reply packages/edit-post/src/editor.js]: And I agree that we need to find a better stable name here.

Comment (michalczaplinski) [reply packages/edit-post/src/editor.js]: > this is passed down from "edit-post" rather than just added automatically from use-block-editor-settings.js

There was no special reason 🙂 It seemed more natural to add it there, but I wasn't 100% sure about it to be honest. 

If we add it in `use-block-editor-settings.js` in the `editor` package, the same setting be available in both the post editor and the site editor, right?

File: packages/editor/src/components/provider/use-block-editor-settings.js [hunk 2]
```diff
@@ -237,6 +241,7 @@ function useBlockEditorSettings( settings, postType, postId ) {
 					BLOCK_EDITOR_SETTINGS.includes( key )
 				)
 			),
+			__experimentalStyles: styles,
```
Comment (michalczaplinski) [inline packages/editor/src/components/provider/use-block-editor-settings.js]: The settings already contain a `styles` key (it's in the `settings` object [above](https://github.com/WordPress/gutenberg/pull/59929/files#diff-278a4a275170c4215a2f4e9d5c0c6d831f9328258220a76cda04b8d9f39f01a5R240))

This `styles` object contains the "compiled" styles, (the CSS) which is not what we want here. What we want is the styles object that contains the user styles.

Another alternative to adding a new `__experimentalStyles` key might be to nest both the "compiled" and "uncompiled" styles under `settings.styles`:

- `settings.styles.raw` - the "uncompiled" styles (what we're adding here)
- `settings.styles.css` - the styles currently present under `settings.styles`

The naming is of course debatable, I'm only suggesting what the structure might be.

File: packages/edit-post/src/editor.js (hunk 1 above)
Comment (youknowriad) [reply packages/edit-post/src/editor.js]: > If we add it in use-block-editor-settings.js in the editor package, the same setting be available in both the post editor and the site editor, right?

yes

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (youknowriad) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: Yes, I think the main blocker of this PR is the naming. It's also important in the discussion to consider that we also have `__experimentalFeatures` that corresponds to the "settings" part of global styles and this new key corresponds to the "styles" part.

The other thing to note is that some of the "compiled" styles mention above are a duplication of the "raw" styles but not all of them. Some other styles come from CSS files themes or plugins inject.

Comment (youknowriad) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: @WordPress/gutenberg-core We need your help coming up with the best API here for the block editor package (generic package to build block editors). How should we pass the "global styles" styles and settings and what about "extra styles" that can be provided as raw CSS to be applied to the editor canvas.

Comment (draganescu) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: At 1st sight I like the idea to split styles in two:

```
styles.source
styles.compiled
```

I think the concept of `raw` doesn't apply since we usually take something raw and transform it into a prepared version. In this case tho we compile to CSS some JSON data.

Comment (draganescu) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: > What we want is the styles object that contains the user styles.

Why can't we then call the key `userStyles`?

File: packages/edit-post/src/editor.js (hunk 1 above)
Comment (michalczaplinski) [reply packages/edit-post/src/editor.js]: makes sense. I moved it to the the `editor` package now 👍

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (michalczaplinski) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: We probably can! I am not yet 100% sure that we won't need ALL (theme and user-defined) styles in the near future as I discover what's needed for the implementation. That's the only reason why I didn't name it `userStyles` 🙂

**Comment by ramonjd** [comment]
Thanks for working on this!! _I think_ this has been tried (and abandoned) before:

- https://github.com/WordPress/gutenberg/pull/34178
- https://github.com/WordPress/gutenberg/pull/34156

So very happy to see it revived!

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (youknowriad) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: Noting that we can't split the existing "styles" setting for backward compatibility reasons.

**Comment by aaronrobertshaw** [comment]
Another +1 for this. Access to the global styles data is also required for the enhanced block style variations feature:
- https://github.com/WordPress/gutenberg/pull/57908

**Comment by michalczaplinski** [comment]
Thanks for the links, @ramonjd! 

I didn't get to review the previous PRs before my AFK so will have to get back to it the week after the next.

**Comment by ramonjd** [comment]
> I didn't get to review the previous PRs before my AFK so will have to get back to it the week after the next.

No worries, I was just dumping them here for cross-linking purposes. Hopefully it will help future Github archaeologists 😄

**Comment by youknowriad** [comment]
I've been thinking about this, the main blocker is that we're not able to decide on a name for this setting. Can we start by using a private setting then? We already have a `ExperimentalBlockEditorProvider` so it shouldn't be too hard to add and unblock this PR.

File: packages/editor/src/components/use-styles/index.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import { useMemo } from '@wordpress/element';
+import { useSelect } from '@wordpress/data';
+import { store as coreStore } from '@wordpress/core-data';
+
+export function useStyles() {
+	const { styles, isReady } = useSelect( ( select ) => {
+		const { getEditedEntityRecord, hasFinishedResolution } =
+			select( coreStore );
+		const _globalStylesId =
+			select( coreStore ).__experimentalGetCurrentGlobalStylesId();
```
Comment (tyxla) [inline packages/editor/src/components/use-styles/index.js]: Should we combine those 2 `select( coreStore )` into a single one?

File: packages/editor/src/components/use-styles/index.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		return {
+			isReady: hasResolved,
+			styles: record?.styles,
+		};
+	}, [] );
+
+	// Make sure to recompute the styles when hasResolved changes.
+	const config = useMemo( () => {
+		return {
+			styles: styles ?? {},
```
Comment (tyxla) [inline packages/editor/src/components/use-styles/index.js]: Should this always default to the same empty object rather than defining a new one on every computation?

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (ramonjd) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: > Noting that we can't split the existing "styles" setting for backward compatibility reasons.

It would be neat to "reclaim" styles for the global styles object, either by splitting as mentioned above or otherwise. 

Blocks use "styles" for the style object and so does theme.json. 

Maybe putting compiled CSS under `"css"` 🤷🏻 

Could we do some sort of type check for backwards compat? E.g.,

```js
// Assume current structure of `[ { ... }, { ... } ]` 
// If we want to be sure, `styles?.[0].hasProperty( isGlobalStyles )`
if ( Array.isArray( styles ) ) { ... }
```

**Comment by noisysocks** [comment]
Is there a simple feature or an enhancement we can include as part of this PR that uses `useStyle`? I'm struggling to review this because it's hard to judge whether or not the hook is appropriate for our needs when it lives in a vacuum.

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (noisysocks) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: Maybe rename this new attribute to `globalStyles` and rename the existing `styles` attribute to `css`?

**Comment by noisysocks** [comment]
How does this new hook relate to `useGlobalStyle` in `@wordpress/block-editor` (which accesses styles via context)? Would it replace that eventually or would they both exist? I'm a bit lost about how this fits into any "grand plan" 😀

**Comment by talldan** [comment]
@noisysocks There's a chance it might be used in #57908, as that requires the uncompiled styles atm.

Context in this discussion thread - https://github.com/WordPress/gutenberg/pull/57908#discussion_r1560199106.

Still investigating whether it's the right fit though and what's left to do.

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (talldan) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: > rename the existing styles attribute to css?

Wouldn't that break back compat?

**Comment by aaronrobertshaw** [comment]
> There's a chance it might be used in https://github.com/WordPress/gutenberg/pull/57908, as that requires the uncompiled styles atm.

This is correct 👍 

FWIW I've taken this PR's approach for a quick spin alongside #57908 and it works well so far. 

If I understand the discussion to date, the main [sticking point](https://github.com/WordPress/gutenberg/pull/59929#discussion_r1531157917) is just the key this styles data will live under?

> How does this new hook relate to useGlobalStyle in @wordpress/block-editor (which accesses styles via context)? Would it replace that eventually or would they both exist? I'm a bit lost about how this fits into any "grand plan" 😀

This is a good question. Should the global styles data always be available in the editor or only if the editor has that context? 

If it is the latter, #57908 would require the post editor set up the GlobalStylesContext provider as per the site editor. I suspect the style inheritance work would have the same need too.

File: packages/editor/src/components/provider/use-block-editor-settings.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: +1 for `globalStyles` 

It covers the potential to include theme and user-defined styles as noted above.

Comment (noisysocks) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: > > rename the existing styles attribute to css?
> 
> Wouldn't that break back compat?

I should have specified "rename in a backwards compatible way" :) so support both but prefer `css`. Could also look at `deprecate()`ing `styles` but it might be too noisy.

Comment (ramonjd) [reply packages/editor/src/components/provider/use-block-editor-settings.js]: Good discussion!

+1 for renaming/repurposing styles - globalStyles is as good as any name. Will there ever be `customStyles` or `userStyles`, or do we assume that global styles will be theme + user merged?

Even if it's noisy for a while to rename/deprecate, I don't think backwards compatibility should be so stultifying a mandate that it should get in the way of progress! Onwards! 🤣

File: packages/editor/src/components/use-styles/index.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		}
+
+		return {
+			isReady: hasResolved,
+			styles: record?.styles,
+		};
+	}, [] );
+
+	// Make sure to recompute the styles when hasResolved changes.
+	const config = useMemo( () => {
```
Comment (ramonjd) [inline packages/editor/src/components/use-styles/index.js]: I was wondering whether we need `useMemo` on the return value of `useSelect` here since I assumed `useSelect` does the shallow comparison on the return values?

cc @jsnajdr for head check.

Comment (jsnajdr) [reply packages/editor/src/components/use-styles/index.js]: The function of this `useMemo` is to keep an unchanged `config` object when its fields, `styles` and `isReady` are unchanged. So that `useStyles` doesn't return a different object on each call.

But the way how `useStyles` is used makes this unnecessary. The return value is destructured:
```js
const { styles } = useStyles();
```
so the `config` object disappears, it's not used anywhere as a prop or an effect dependency.

There is this comment above the `useMemo` call:
> Make sure to recompute the styles when hasResolved changes.

It's confusing. The `styles` are fully computed, the hook is not recomputing anything. It's just ensuring that referentially-identical input leads to referentially-identical output.

**Comment by aaronrobertshaw** [comment]
I've rebased this PR to resolve a few conflicts and prepare for merging https://github.com/WordPress/gutenberg/pull/61022 back into this one.

**Comment by aaronrobertshaw** [comment]
I've merged into this branch #61022 that addresses most of the feedback received either here on on that PR. It also updates the approach slightly so that `useGlobalStylesData` remains private for now and the global styles data includes the merged data from all origins e.g. base, user etc.

File: packages/editor/src/components/use-global-styles-data/index.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	const [ isUserStylesReady, userStyles ] = useGlobalStylesUserData();
+	const mergedStyles = useMemo(
+		() => mergeBaseAndUserStyles( baseStyles, userStyles ),
+		[ baseStyles, userStyles ]
+	);
+
+	return {
+		isReady: isBaseStylesReady && isUserStylesReady,
+		base: baseStyles ?? DEFAULT_STYLES,
+		user: userStyles ?? DEFAULT_STYLES,
```
Comment (scruffian) [inline packages/editor/src/components/use-global-styles-data/index.js]: Is it a good idea to return base, user and merged, or should we just return merged?

Comment (aaronrobertshaw) [reply packages/editor/src/components/use-global-styles-data/index.js]: I did wonder about this myself.

It could be argued returning the different origins' data is more future-proof. For example, the style inheritance UI could differ based on at which level the style was modified. 

The current approach also reflects the data in the site editor's global styles context providing more consistency. Which could help consolidate some of this down the road.

Comment (youknowriad) [reply packages/editor/src/components/use-global-styles-data/index.js]: I'm not entirely sure here. If one think about the block editor package as a standalone package, it doesn't really know that there are multiple levels of style configurations and probably doesn't matter for blocks where the style is coming from. I'd love to have more details about the potential UIs before committing to this.

Comment (aaronrobertshaw) [reply packages/editor/src/components/use-global-styles-data/index.js]: > think about the block editor package as a standalone package, it doesn't really know that there are multiple levels of style configurations

That's a good point, thanks @youknowriad 👍 

One other option I was pondering was only adding the `merged` property for now keeping the door open to add the others later if needed.

To your point though maybe I should switch this PR to only return the merged values.

@richtabor can probably confirm the style inheritance UI work only needs the merged data.

Comment (richtabor) [reply packages/editor/src/components/use-global-styles-data/index.js]: >  I'd love to have more details about the potential UIs before committing to this.

I don't think that we need much (perhaps any) UI for style inheritance. If we can get the default values available on blocks, and leverage the existing reset/clear UI. 

For example, this could be an unedited group block, with its styles derived from global styles, with a font size of XL applied locally: 

CleanShot 2024-05-09 at 17 02 59

I'd expect resetting the font size would set it to whatever the current global styles value is (regardless of if its `theme` or `user`).

Comment (aaronrobertshaw) [reply packages/editor/src/components/use-global-styles-data/index.js]: So it looks like we only have a use case for the merged data so far. I'll get this updated to only add the merged styles to the block editor settings.

Comment (aaronrobertshaw) [reply packages/editor/src/components/use-global-styles-data/index.js]: I've updated this PR to return only the merged styles data. Let me know if there is anything else that needs tweaking.

Comment (aaronrobertshaw) [reply packages/editor/src/components/use-global-styles-data/index.js]: Rethinking this a bit, I'd like to reduce some code duplication with the site editor's global styles context setup. 

This would require the user, base, and merged configs to be exported from a new hook in the editor package. The editor setting however will still only be the merged style data as discussed.

I'll update again when this is ready for another look.

Comment (aaronrobertshaw) [reply packages/editor/src/components/use-global-styles-data/index.js]: There's an alternate approach to this PR up in https://github.com/WordPress/gutenberg/pull/61556.

TL;DR

The new approach relocates the collection of GlobalStylesContext data to the editor package. This way we have a single location that collects and merges the global styles data. So when block style variations need to be resolved within that merged data, it only happens once and is available in both editors.

After moving the GlobalStylesContext it seemed to make sense that the GlobalStylesProvider was also in the editor package. Happy to move if required.

**Comment by scruffian** [comment]
Closing because https://github.com/WordPress/gutenberg/pull/61556 is merged
